internal/route/api/v1/repo/path.go

```diff
@@ -0,0 +1,198 @@
+	Encoding        string `json:"encoding,omitempty"`
+}
+
+type Links struct {
+	Git  string `json:"git"`
+	Self string `json:"self"`
+	HTML string `json:"html"`
+}
+
+const (
+	// :base-url/:username/:project/raw/:refs/:path
+	templateDownloadURL = "%s/%s/%s/raw/%s"
+	// :base-url/repos/:username/:project/contents/:path
+	templateSelfLink = "%s/repos/%s/%s/contents/%s"
+	// :baseurl/repos/:username/:project/git/trees/:sha
+	templateGitURLLink = "%s/repos/%s/%s/trees/%s"
+	// :baseurl/repos/:username/:project/tree/:sha
+	templateHTMLLLink = "%s/repos/%s/%s/tree/%s"
+)
+
+func GetPathContents(c *context.APIContext) {
+	treeEntry, err := c.Repo.Commit.GetTreeEntryByPath(c.Repo.TreePath)
+	if err != nil {
+		c.NotFoundOrServerError("GetTreeEntryByPath", git.IsErrNotExist, err)
+	}
+
+	gitURL := fmt.Sprintf(templateGitURLLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), treeEntry.ID.String())
+	htmlURL := fmt.Sprintf(templateHTMLLLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), treeEntry.ID.String())
+	selfurl := fmt.Sprintf(templateSelfLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), c.Repo.TreePath)
+
+	gr := &GitRepoContent{
+		DownloadURL: fmt.Sprintf(templateDownloadURL, c.BaseURL, c.Params(":username"), c.Params(":reponame"), c.Repo.TreePath),
+		Size:        treeEntry.Size(),
+		Name:        treeEntry.Name(),
+		Path:        c.Repo.TreePath,
+		Sha:         treeEntry.ID.String(),
+		Links: Links{
+			Git:  gitURL,
+			Self: selfurl,
+			HTML: htmlURL,
+		},
+		URL:     selfurl,
+		GitURL:  gitURL,
+		HTMLURL: htmlURL,
+	}
+	// A tree entry can only be one of the following types:
+	// 1. Tree (Directory)
+	// 2. SubModule
+	// 3. SymLink
+	// 4. Blob
+	if treeEntry.IsDir() {
+		dirTree, err := c.Repo.GitRepo.GetTree(treeEntry.ID.String())
+		if err != nil {
+			c.NotFoundOrServerError("GetGitDirTree", git.IsErrNotExist, err)
+			return
+		}
+
+		entries, err := dirTree.ListEntries()
+		if err != nil {
+			c.NotFoundOrServerError("ListDirTreeEntries", git.IsErrNotExist, err)
+			return
+		}
+
+		results, err := AppendDirTreeEntries(entries, c)
+
+		if err != nil {
+			c.NotFoundOrServerError("AppendDirTreeEntries", git.IsErrNotExist, err)
+			return
+
+		}
+		c.JSONSuccess(results)
+		return
+
+	} else if treeEntry.IsSubModule() {
+		gr.Type = "submodule"
+		parsedURL, err := url.Parse(c.BaseURL)
+		if err != nil {
+			c.ServerError("ErrorURLParse", err)
+		}
+		host := parsedURL.Host
+		submoduleURL := fmt.Sprintf("git://%s/%s/%s.git", host, c.Params(":name"), c.Params(":reponame"))
+		gr.SubmoduleGitURL = submoduleURL
+		c.JSONSuccess(gr)
+		return
+
+	} else if treeEntry.IsLink() {
+		gr.Type = "symlink"
+		gr.Target = c.Repo.TreePath
+		c.JSONSuccess(gr)
+		return
+
+	} else { // tree entry is a blob
+		gr.Type = "blob"
+		b, err := getBase64EncodedBlob(c)
+
+		if err != nil {
+			c.ServerError("GetBlobContent", err)
+			return
+		}
+
+		gr.Content = b
+		c.JSONSuccess(gr)
+		return
+	}
```

Comment:
We can check these three cases (with shorter code logic) first, then deal with the `treeEntry.IsDir()` case to save one level of indentation, since that chuck of code is large.

Reply:
Sure makes sense to me. This should be fixed now.